newIDE/app/src/CommandPalette/CommandPalette/index.js

```diff
@@ -19,76 +20,115 @@ const useStyles = makeStyles({
+  (props, ref) => {
+    const classes = useStyles();
+    const commandManager = React.useContext(CommandsContext);
+    const [mode, setMode] = React.useState<PaletteMode>('closed');
+    const [
+      selectedCommand,
+      selectCommand,
+    ] = React.useState<null | NamedCommandWithOptions>(null);
 
-  const handleOptionChoose = (option: CommandOption) => {
-    option.handler();
-    props.onClose();
-  };
+    // Takes a command and if simple command, executes handler
+    // If command with options, opens options of the palette
+    const handleCommandChoose = React.useCallback(
+      (command: NamedCommand) => {
+        if (command.handler) {
+          // Simple command
+          command.handler();
+          if (command.name !== 'OPEN_COMMAND_PALETTE') {
+            // Don't close palette if the command is for opening it
+            setMode('closed');
+          }
+        } else {
+          // Command with options
+          selectCommand(command);
+          setMode('option');
+        }
+      },
+      [selectCommand, setMode]
+    );
 
-  return (
-    <I18n>
-      {({ i18n }) => (
-        <Dialog
-          onClose={props.onClose}
-          aria-label="command-palette"
-          open={props.open}
-          fullWidth
-          hideBackdrop
-          maxWidth="sm"
-          classes={classes}
-          transitionDuration={0}
-        >
-          {commandPickerOpen && (
-            // Command picker
-            <AutocompletePicker
-              i18n={i18n}
-              items={commandManager.getAllNamedCommands()}
-              placeholder={t`Start typing a command...`}
-              onClose={props.onClose}
-              onSelect={handleCommandChoose}
-            />
-          )}
-          {optionPickerOpen && selectedCommand && (
-            // Command options picker
-            <AutocompletePicker
-              i18n={i18n}
-              items={selectedCommand.generateOptions()}
-              placeholder={selectedCommand.displayText}
-              onClose={props.onClose}
-              onSelect={handleOptionChoose}
-            />
-          )}
-        </Dialog>
-      )}
-    </I18n>
-  );
-};
+    // Executes handler of a command option and closes palette
+    const handleOptionChoose = (option: CommandOption) => {
+      option.handler();
+      setMode('closed');
+    };
+
+    // Opens the palette in command mode
+    const openPalette = React.useCallback((open? = true) => {
+      if (open) setMode('command');
+      else setMode('closed');
+    }, []);
+
+    // Takes command name, gets command object from
+    // manager and launches command accordingly
+    const launchCommand = React.useCallback(
+      commandName => {
+        const command = commandManager.commands[commandName];
+        if (!command) return;
+        handleCommandChoose({ ...command, name: commandName });
+      },
+      [handleCommandChoose, commandManager.commands]
+    );
```

Comment:
Could we remove commands from the interface entirely? Otherwise a `getCommand` is surely slightly better - because it's left to the implementor to choose how the `getCommand` should work. Safer compared to forcing an internal detail like `commands` field to be visible :)